File: python/binary_search_tree.py
# Binary Search Tree is a binary tree with following properties for each node in the tree:
# 1. The left subtree has keys less than the current node
# 2. The right subtree has keys more than the current node
# 3. There are no duplicate nodes.

class Node:

    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None


def search(root, key):
    if not root or root.val == key:
        return root

    if root.val < key:
        return search(root.right, key)

    return search(root.left, key)


def insert(root, key):
    if not root:
        return Node(key)

    if root.val < key:
        root.right = insert(root.right, key)
    elif root.val > key:
        root.left = insert(root.left, key)

    return root


def find_successor(root):
    curr = root.right

    while(curr.left):
        curr = curr.left

    return curr


def delete(root, key):
    if not root:
        return root

    if root.val < key:
        root.right = delete(root.right, key)
    elif root.val > key:
        root.left = delete(root.left, key)
    else:
        if not root.left and not root.right:
            return None
        elif not root.left:
            temp = root.right
            root.right = None
            return temp

        elif not root.right:
            temp = root.left
            root.left = None
            return temp
        else:
            successor = find_successor(root)
            root.val = successor.val
            delete(root.right, successor.val)
            return root


if __name__ == '__main__':
    r = Node(50)
    r = insert(r, 30)
    r = insert(r, 20)
    r = insert(r, 40)
    r = insert(r, 70)
    r = insert(r, 60)
    r = insert(r, 80)
    print(search(r, 46) == None)
    print(search(r, 60) == None)
    print(delete(r, 70))

File: go/binary_search_tree.go
package main

import "fmt"

type Node struct {
	value int
	left  *Node
	right *Node
}

type BinarySearchTree struct {
	root *Node
}

func search(root *Node, key int) *Node {
	if root == nil || root.value == key {
		return root
	}

	if root.value < key {
		return search(root.right, key)
	}

	return search(root.left, key)
}

func insert(root *Node, key int) *Node {
	if root == nil {
		return &Node{
			value: key,
		}
	}

	if root.value < key {
		root.right = insert(root.right, key)
	} else {
		root.left = insert(root.left, key)
	}

	return root
}

func findSuccessor(root *Node) *Node {
	curr := root.right

	for curr.left != nil {
		curr = curr.left
	}

	return curr
}

func delete(root *Node, key int) *Node {
	if root == nil {
		return root
	} else if root.value < key {
		root.right = delete(root.right, key)
	} else if root.value > key {
		root.left = delete(root.left, key)
	} else {
		if root.left == nil && root.right == nil {
			return nil
		} else if root.left == nil {
			return root.right
		} else if root.right == nil {
			return root.left
		} else {
			successor := findSuccessor(root)
			root.value = successor.value
			delete(root.right, successor.value)
			return root
		}
	}

	return nil
}

func main() {

	root := &Node{
		value: 50,
	}

	binarySearchTree := &BinarySearchTree{
		root: root,
	}

	root = insert(binarySearchTree.root, 30)
	root = insert(binarySearchTree.root, 20)
	root = insert(binarySearchTree.root, 40)
	root = insert(binarySearchTree.root, 70)
	root = insert(binarySearchTree.root, 60)
	root = insert(binarySearchTree.root, 80)

	fmt.Println(search(binarySearchTree.root, 46) != nil)
	fmt.Println(search(binarySearchTree.root, 60) != nil)

}
